db/system_keyspace.hh

```diff
@@ -364,6 +367,48 @@ public:
     future<replay_positions> get_truncated_positions(table_id);
     future<> drop_truncation_rp_records();
 
+    // When a commitlog replay happens after a successful cleanup operation,
+    // we have to filter out the mutations affected by the cleanup,
+    // to avoid data resurrection.
+    //
+    // For this purpose, records of cleanup operations (the affected token ranges
+    // and commitlog ranges) are kept in a system table.
+    //
+    // The below functions manipulate these records. 
+    future<> save_range_cleanup_record(table_id, dht::token_range, db::replay_position);
+    struct cleanup_map_hash {
+        size_t operator()(const std::pair<table_id, int32_t>& p) const {
+            size_t seed = 0;
+            boost::hash_combine(seed, std::hash<utils::UUID>()(p.first.uuid()));
+            boost::hash_combine(seed, std::hash<int32_t>()(p.second));
+            return seed;
+        }
+    };
+    // For a given token, this map returns the minimum replay position unaffected by cleanups.
+    // A mutation in commitlog should only be replayed if it lies on a replay position
+    // greater or equal to that minimum for its token.
```

Comment:
Need to carefully test off-by-ones here, not easy.

Reply:
I think the tests I added are doing a decent job at testing the off-by-ones. (In particular, I tried flipping between `<` and `<=` in the places I could think of and the tests caught that).